refactor: use ngrx store for user tags

tamazlykar opened the pull request:
This PR implement the root store with a user-tags part. 
PR fixes #1014 #1140 #1143 #1142 
Additionally fixed bugs:
- There was a bug with agreement dialog during VM creation, because of the using deprecated functionality of angular - fixed.
- If you enable autosave password for VMs and create new VM that does not have a password then you will have a VM password tag = undefined. Now if VM does not have a password a tag will not create.

Additional improvements:
- Added error notification if the passwords do not match in the password change form. The error appears only when both fields are filled.
- Added a dialog asking confirmation about password update (in the same way when you want to generate API keys).
- Added a dialog notification if API keys regeneration produces an error.
- Added a snackbar notification (currently for user-tags only) that notify when there was an error about data loading. This snackbar has a refresh button that refreshes the whole app. (see screenshot below)

Additional info:
- There is a bug that when you open the VM creation agreement dialog scrolls this dialog to the end.
**Update:** test was fixed

![screenshot from 2018-07-10 13-40-48](https://user-images.githubusercontent.com/18456914/42609370-77778858-85b6-11e8-940c-aee5dfb1b9a2.png)

tamazlykar commented:
Is it a good decision to separate a settings effects and settings view model selectors to the settings module or it will be better to place inside user-tags store effects and selectors?

zolotyx commented on src/app/home/user-settings-resolver.service.ts:
```diff
@@ -0,0 +1,21 @@
+    this.store.dispatch(new UserTagsActions.LoadUserTags());
+
+    return this.store.select(UserTagsSelectors.getIsLoading)
+      .filter(loading => loading === false)
```
.filter(loading => !loading)

zolotyx commented on src/app/root-store/user-tags/user-tags.reducer.ts:
```diff
@@ -0,0 +1,55 @@
+  const upsertEntities = getEntities(entities);
+  return {
+    ...state,
+    ids: [...state.ids, ...upsertIds],
```
could there be some duplicate ids?

tamazlykar replied:
You right, it could be. Changed with ngrx/entities.

zolotyx commented on src/app/root-store/user-tags/user-tags.reducer.ts:
```diff
@@ -0,0 +1,55 @@
+  return entities.map(tag => tag.key)
+}
+
+function upsertMany(entities: Tag[], state: UserTagsState): UserTagsState {
```
Why not to use createEntityAdapter from ngrx/entity??

tamazlykar replied:
Because it works only with ngrx/entity itself. I did not use it because I can't prepopulate store(ids, entities) during initialization.

tamazlykar replied:
I was wrong, it's possible. Changed.

tamazlykar commented:
I fix the tests for IDLE Monitor functionality.

zolotyx commented on src/app/settings/store/settings.effects.ts:
```diff
@@ -0,0 +1,77 @@
+  value: any
+}
+
+function getDifferentKeysAndValues(current: SettingsViewModel, changed: SettingsViewModel): Change[] {
```
think about using some lodash method

tamazlykar replied:
Can't find any lodash method that will simplify this part of the code. Instead, I added tests for this function.

zolotyx commented on src/app/settings/store/settings.effects.ts:
```diff
@@ -0,0 +1,77 @@
+
+  private dispatchUpdateSetting(change: Change) {
+    const value = change.value;
+    switch (change.key) {
```
no default case

tamazlykar replied:
Added

zolotyx commented on src/app/shared/classes/app-configuration.ts:
```diff
@@ -0,0 +1,55 @@
+import * as defaultConf from '../../../config/default-config.json';
```
json configs should not be included in the code

zolotyx commented on src/app/vm/shared/vm.service.ts:
```diff
@@ -34,28 +26,23 @@ export class VmService extends BaseBackendService<VirtualMachine> {
-        return this.getListWithDetails(undefined, true)
-          .switchMap(vmList => this.userTagService.setLastVmId(vmList.length));
-      });
+    return this.store.select(UserTagsSelectors.getLastVMId).first();
```
do not use store in services

tamazlykar replied:
ok

zolotyx commented on src/app/tags/tags-view/tags-view.component.ts:
```diff
@@ -162,7 +162,7 @@ export class TagsViewComponent implements OnInit, OnChanges {
   private getCategories(): Array<TagCategory> {
     const groupedTags = groupBy(
-      this.tags.map(_ => Object.assign({}, _, {categoryName: categoryName(_)})),
+      this.tags.map(_ => Object.assign({}, _, { categoryName: categoryName(_) })),
```
replace _, use spread

tamazlykar replied:
changed

tamazlykar commented on src/app/root-store/user-tags/user-tags.selectors.ts:
```diff
@@ -0,0 +1,95 @@
+import { userTagKeys } from './user-tag-keys';
+import { DayOfWeek, Language, TimeFormat } from '../../shared/types';
+
+function convertToBoolean(input: string): boolean | undefined {
```
Yes, I think it will be better

tamazlykar commented on src/app/root-store/user-tags/user-tags.selectors.ts:
```diff
@@ -0,0 +1,95 @@
+
+export const getNavigationOrder = createSelector(
+  getUserTagsEntities,
+  (entities): string | null => entities[userTagKeys.navigationOrder].value
```
It was decided to use an empty string instead of null

tamazlykar commented on src/app/settings/containers/settings.component.ts:
```diff
@@ -0,0 +1,127 @@
+import { Component } from '@angular/core';
+import { Store } from '@ngrx/store';
+import { Observable } from 'rxjs/Observable';
+import 'rxjs/add/operator/first';
```
fixed

tamazlykar commented on src/app/settings/directives/settings-section-content.directive.ts:
```diff
@@ -0,0 +1,9 @@
+@Directive({ selector: '[csSettingsSectionContent]' })
+export class SettingsSectionContentDirective {
+  constructor(private el: ElementRef) {
+    el.nativeElement.style.marginLeft = '20px';
```
Changed to a shared style file

tamazlykar commented on src/app/settings/directives/settings-section-name.directive.ts:
```diff
@@ -0,0 +1,9 @@
+@Directive({ selector: '[csSettingsSectionName]' })
+export class SettingsSectionNameDirective {
+  constructor(private el: ElementRef) {
+    el.nativeElement.style.fontSize = '22px';
```
Changed to a shared style file

tamazlykar commented on src/app/settings/containers/settings.component.ts:
```diff
@@ -0,0 +1,127 @@
+export class SettingsComponent {
+  public settings$: Observable<SettingsViewModel>;
+  public userKeys: ApiKeys;
+  public apiUrl = 'http';
```
+

tamazlykar commented:
Rewritten partially store for tags. Added actions for each system tag. Inspired by [this video](https://www.youtube.com/watch?v=JmnsEvoy-gY).

The configuration of the application has been rewritten. The system has an internal configuration file that merges with the user configuration file. Now the system does not have the conf.json file by default. In another PR, conf-example.json will be added.

For the following configuration parameters, validation exists:
- defaultDomain - check for a string
- sessionRefreshInterval - check for a number and a non-negative value
- defaultFirstDayOfWeek - on a number from 0 to 1
- defaultInterfaceLanguage - valid parameters 'en', 'ru'
- defaultTimeFormat - valid parameters for 'hour12', 'hour24', 'auto'
- defaultThemeName - valid parameters 'blue-red', 'indigo-pink'
- sessionTimeout - check for a number and a non-negative value

The system displays a warning about an incorrect configuration in the console when:
- if the parameter specified in the config does not exist
- if the parameter has an invalid value (only for parameters with validation)
- if the config is not a json object

zolotyx commented on .gitignore:
```diff
@@ -58,5 +58,6 @@ __build__/**
-# Proxy config #
+# Configurations #
 proxy-conf.js
+/src/config/config.json
```
maybe ignore the whole directory?

tamazlykar replied:
The directory contains the timezones.json file, so we don't need to ignore whole directory

zolotyx commented on src/app/core/services/system-tags.service.ts:
```diff
@@ -0,0 +1,66 @@
+import { Tag } from '../../shared/models';
+import { userTagKeys } from '../../tags/tag-keys';
+
+function toString(value: boolean | number): string {
```
do you need this function?

zolotyx commented on src/app/root-store/idle-monitor/idle-monitor.effects.ts:
```diff
@@ -0,0 +1,98 @@
+  );
+
+  @Effect({ dispatch: false })
+  updateIdleMonitorTimeout$: Observable<number> = this.actions$.pipe(
```
I'd merge 2 effects. this one and the previous

zolotyx commented on src/app/root-store/server-data/user-tags/user-tags.effects.ts:
```diff
@@ -0,0 +1,318 @@
+const castValueToString = (value: any): string => {
+  switch (typeof value) {
+    case 'boolean': {
+      return value ? 'true' : 'false';
```
return `${value}`;

zolotyx commented on src/app/root-store/server-data/user-tags/user-tags.effects.ts:
```diff
@@ -0,0 +1,318 @@
+import * as userTagsSelectors from './user-tags.selectors';
+import { StartIdleMonitor, UpdateIdleMonitorTimeout } from '../../idle-monitor/idle-monitor.actions';
+
+const castValueToString = (value: any): string => {
```
not sure if this function is required

tamazlykar commented:
run tests
